pkg/controller/vmnetcfg/controller.go

```diff
@@ -136,40 +146,29 @@ func (h *Handler) Allocate(vmNetCfg *networkv1.VirtualMachineNetworkConfig, stat
 			)
 
 			// Update IPPool status
-			ipPoolNamespace, ipPoolName := kv.RSplit(nc.NetworkName, "/")
-			if err := retry.RetryOnConflict(retry.DefaultBackoff, func() error {
-				ipPool, err := h.ippoolCache.Get(ipPoolNamespace, ipPoolName)
-				if err != nil {
-					return err
-				}
-
-				ipPoolCpy := ipPool.DeepCopy()
+			ipPoolCpy := ipPool.DeepCopy()
 
-				ipv4Status := ipPoolCpy.Status.IPv4
-				if ipv4Status == nil {
-					ipv4Status = new(networkv1.IPv4Status)
-				}
+			ipv4Status := ipPoolCpy.Status.IPv4
+			if ipv4Status == nil {
+				ipv4Status = new(networkv1.IPv4Status)
+			}
 
-				allocated := ipv4Status.Allocated
-				if allocated == nil {
-					allocated = make(map[string]string)
-				}
+			allocated := ipv4Status.Allocated
+			if allocated == nil {
+				allocated = make(map[string]string)
+			}
 
-				allocated[ip] = nc.MACAddress
+			allocated[ip] = nc.MACAddress
 
-				ipv4Status.Allocated = allocated
-				ipPoolCpy.Status.IPv4 = ipv4Status
+			ipv4Status.Allocated = allocated
+			ipPoolCpy.Status.IPv4 = ipv4Status
 
-				if !reflect.DeepEqual(ipPoolCpy, ipPool) {
-					logrus.Infof("(vmnetcfg.Allocate) update ippool %s/%s", ipPool.Namespace, ipPool.Name)
-					ipPoolCpy.Status.LastUpdate = metav1.Now()
-					_, err = h.ippoolClient.UpdateStatus(ipPoolCpy)
-					return err
+			if !reflect.DeepEqual(ipPoolCpy, ipPool) {
+				logrus.Infof("(vmnetcfg.Allocate) update ippool %s/%s", ipPool.Namespace, ipPool.Name)
```

Comment:
I think those logic are similar, could it be like that?

```go
func (h *Handler) Allocate(vmNetCfg *networkv1.VirtualMachineNetworkConfig, status networkv1.VirtualMachineNetworkConfigStatus) (networkv1.VirtualMachineNetworkConfigStatus, error) {
	logrus.Debugf("(vmnetcfg.Allocate) allocate ip for vmnetcfg %s/%s", vmNetCfg.Namespace, vmNetCfg.Name)

	if vmNetCfg.Spec.Paused != nil && *vmNetCfg.Spec.Paused {
		return status, fmt.Errorf("vmnetcfg %s/%s was administratively disabled", vmNetCfg.Namespace, vmNetCfg.Name)
	}

	var ncStatuses []networkv1.NetworkConfigStatus

	for _, nc := range vmNetCfg.Spec.NetworkConfig {
		ipPoolNamespace, ipPoolName := kv.RSplit(nc.NetworkName, "/")
		ipPool, err := h.ippoolCache.Get(ipPoolNamespace, ipPoolName)
		if err != nil {
			return status, err
		}

		if !networkv1.CacheReady.IsTrue(ipPool) {
			return status, fmt.Errorf("ippool %s/%s is not ready", ipPoolNamespace, ipPoolName)
		}

		exists, err := h.cacheAllocator.HasMAC(nc.NetworkName, nc.MACAddress)
		if err != nil {
			return status, err
		}

		var ncStatus networkv1.NetworkConfigStatus

		if exists {
			// Recover IP from cache
			ip, err := h.cacheAllocator.GetIPByMAC(nc.NetworkName, nc.MACAddress)
			if err != nil {
				return status, err
			}

			// Prepare VirtualMachineNetworkConfig status
			ncStatus = networkv1.NetworkConfigStatus{
				AllocatedIPAddress: ip,
				MACAddress:         nc.MACAddress,
				NetworkName:        nc.NetworkName,
				State:              networkv1.AllocatedState,
			}
		} else {
			// Allocate new IP
			dIP := net.IPv4zero.String()
			if nc.IPAddress != nil {
				dIP = *nc.IPAddress
			}
			// Recover IP from status (resume from paused state)
			if oIP, err := findIPAddressFromNetworkConfigStatusByMACAddress(vmNetCfg.Status.NetworkConfig, nc.MACAddress); err == nil {
				dIP = oIP
			}

			ip, err := h.ipAllocator.AllocateIP(nc.NetworkName, dIP)
			if err != nil {
				return status, err
			}

			if err := h.cacheAllocator.AddMAC(nc.NetworkName, nc.MACAddress, ip); err != nil {
				return status, err
			}

			// Prepare VirtualMachineNetworkConfig status
			ncStatus = networkv1.NetworkConfigStatus{
				AllocatedIPAddress: ip,
				MACAddress:         nc.MACAddress,
				NetworkName:        nc.NetworkName,
				State:              networkv1.AllocatedState,
			}
		}
		
		ncStatuses = append(ncStatuses, ncStatus)

		// Update VirtualMachineNetworkConfig metrics
		h.metricsAllocator.UpdateVmNetCfgStatus(
			fmt.Sprintf("%s/%s", vmNetCfg.Namespace, vmNetCfg.Name),
			ncStatus.NetworkName,
			ncStatus.MACAddress,
			ncStatus.AllocatedIPAddress,
			string(ncStatus.State),
		)

		// Update IPPool status
		ipPoolCpy := ipPool.DeepCopy()

		ipv4Status := ipPoolCpy.Status.IPv4
		if ipv4Status == nil {
			ipv4Status = new(networkv1.IPv4Status)
		}

		allocated := ipv4Status.Allocated
		if allocated == nil {
			allocated = make(map[string]string)
		}

		allocated[ip] = nc.MACAddress

		ipv4Status.Allocated = allocated
		ipPoolCpy.Status.IPv4 = ipv4Status

		if !reflect.DeepEqual(ipPoolCpy, ipPool) {
			logrus.Infof("(vmnetcfg.Allocate) update ippool %s/%s", ipPool.Namespace, ipPool.Name)
			ipPoolCpy.Status.LastUpdate = metav1.Now()
			if _, err = h.ippoolClient.UpdateStatus(ipPoolCpy); err != nil {
				return status, err
			}
		}
	}

	status.NetworkConfig = ncStatuses

	return status, nil
}
```

Reply:
Yes, it could be refactored like that. It's more concise now. Thank you.